### PR TITLE
feat: New Graphics API + Flagged WebGL Support

### LINKED ISSUE #136: New Pool Collection Type to support Object Pooling
### Context

Object pooling would bring significant improvements to the memory usage of Excalibur, as well as performance gains related to the construction/destruction of objects.

### Proposal

// Creates a new actor pool
var pool = new ex.Pool<Actor>(100);

// must implement IPoolable
interface IPoolable{
    public reset(): void;
    public dispose(): void;
}

var myActor = pool.getInstance();

// reset and dispose will be called on actor and returned to the pool
pool.returnInstance(myActor);

### PR DESCRIPTION
**Pull request by eonarheim**
===:clipboard: PR Checklist :clipboard:===

- [x] :pushpin: issue exists in github for these changes
- [x] :microscope: existing tests still pass
- [x] :see_no_evil: code conforms to the [style guide](https://github.com/excaliburjs/Excalibur/blob/main/STYLEGUIDE.md)
- [x] :triangular_ruler: new tests written and passing / old tests updated with new scenario(s)
- [ ] :page_facing_up: changelog entry added (or not needed)

==================

Related to v0.25.0 Roadmap #1161

Closes #136 #580 #589 #652 #1529 #420 #172 #276 #384 #387 #481

## TODO before ~~merge~~ next release (team spoke offline)
* [ ] Docs PR
* [ ] New Benchmarks

## Changes:

- Flagged Support for WebGL implementation behind `ex.Flags.enable('use-webgl')`
  - Batched draw calls by default
- New Graphics API
  - `ExcaliburGraphicsContext` as drawing abstraction
  - ECS Style `GraphicsComponent` & `GraphicsSystem`
  - `ImageSource` replaces `Texture`
  - New `Text` Graphics type for easy text drawing
  - New `SpriteFont` and `Font` for manipulating fonts
  - New `Rectangle`, `Circle`, and `Polygon` Graphics
  - New `GraphicsGroup` for grouping Graphics together into one
  - Improved `Animation` Graphic with events on `frame`, `loop`, and `end`
  - New `Canvas` Graphic to act as a `CanvasRenderingContext2D` shim

### CODE REVIEW

File: sandbox/src/game.ts [hunk 1]
```diff
@@ -37,30 +37,39 @@ var logger = ex.Logger.getInstance();
...
+  height: 600 / 2,
+  viewport: { width: 800, height: 600 },
   canvasElementId: 'game',
   suppressHiDPIScaling: false,
-  suppressPlayButton: true
+  suppressPlayButton: true,
+  antialiasing: false,
+  snapToPixel: true
 });
+game.setAntialiasing(false);
```
Comment (kamranayub) [inline sandbox/src/game.ts:51]: Is this needed when using `antialiasing: false`?

File: sandbox/src/game.ts [hunk 2]
```diff
@@ -236,49 +384,69 @@ follower.actions
...
-player.addDrawing(Animations.JumpLeft, jumpLeft);
+player.graphics.add(Animations.Left, left);
+player.graphics.add(Animations.Right, right);
+player.graphics.add(Animations.Idle, idle);
+player.graphics.add(Animations.JumpRight, jumpRight);
+player.graphics.add(Animations.JumpLeft, jumpLeft);
 
 // Set default animation
-player.setDrawing(Animations.Idle);
+player.graphics.swap(Animations.Idle);
```
Comment (kamranayub) [inline sandbox/src/game.ts:449]: Wonder if `.switchTo` or `.use` could be a nicer API?

File: src/engine/Drawing/Sprite.ts [hunk 3]
```diff
@@ -7,12 +7,14 @@ import { Vector } from '../Algebra';
...
 import { Configurable } from '../Configurable';
+import { obsolete } from '../Util/Decorators';
 
 /**
  * @hidden
+ * @deprecated Use [["Graphics/Sprite".Sprite]]
  */
 export class SpriteImpl implements Drawable {
-  private _texture: Texture;
+  public _texture: Texture;
```
Comment (kamranayub) [inline src/engine/Drawing/Sprite.ts:17]:
```suggestion
  public texture: Texture;
```

Changing accessibility maybe need to update name?

File: src/engine/Engine.ts [hunk 4]
```diff
@@ -557,12 +565,31 @@ O|===|* >________________>\n\
       displayMode = DisplayMode.FullScreen;
     }
 
-    // eslint-disable-next-line
-    this.ctx = this.canvas.getContext('2d', { alpha: this.enableCanvasTransparency });
+    if (Flags.isEnabled('use-webgl')) {
```
Comment (kamranayub) [inline src/engine/Engine.ts:568]: May want to introduce flag constants to make it easier to update/change/check.

File: src/engine/EntityComponentSystem/Components/TransformComponent.ts [hunk 5]
```diff
@@ -18,15 +18,16 @@ export enum CoordPlane {
 }
 
 export class TransformComponent extends Component<'transform'> {
+  static type: 'transform';
```
Comment (kamranayub) [inline src/engine/EntityComponentSystem/Components/TransformComponent.ts:21]: Do we need both static and instance property?

File: src/engine/EntityComponentSystem/SystemManager.ts [hunk 6]
```diff
@@ -49,25 +49,30 @@ export class SystemManager<ContextType> {
...
+        s.preupdate(context, delta);
       }
     }
 
     for (const s of systems) {
+      // Get entities that match the system types, pre-sort
       const entities = this._world.queryManager.getQuery(s.types).getEntities(s.sort);
+      // Initialize entities if needed
+      for (const entity of entities) {
+        entity._initialize((context as any)?.engine);// slightly gross
```
Comment (kamranayub) [inline src/engine/EntityComponentSystem/SystemManager.ts:68]: What's happening here? Could we wrap this in a type guard perhaps to make it clearer?

File: src/engine/Graphics/Context/ExcaliburGraphicsContextWebGL.ts [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+import { StateStack } from './state-stack';
+import { Logger } from '../../Util/Log';
+import { LineRenderer } from './line-renderer';
+import { ImageRenderer } from './image-renderer';
+import { PointRenderer } from './point-renderer';
+import { Canvas } from '../Canvas';
+import { GraphicsDiagnostics } from '../GraphicsDiagnostics';
+
+class ExcaliburGraphicsContextWebGLDebug implements DebugDraw {
+  constructor(private _ex: ExcaliburGraphicsContextWebGL) {}
```
Comment (kamranayub) [inline src/engine/Graphics/Context/ExcaliburGraphicsContextWebGL.ts:24]:
```suggestion
  constructor(private _webglCtx: ExcaliburGraphicsContextWebGL) {}
```

File: src/engine/Graphics/Context/ExcaliburGraphicsContextWebGL.ts [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+   * @internal
+   */
+  public __gl: WebGLRenderingContext;
+
+
+  /**
+   * Holds the 2d context shim
+   */
+  private _canvas: Canvas; // Configure z for shim?
+  public __ctx: CanvasRenderingContext2D;
```
Comment (kamranayub) [inline src/engine/Graphics/Context/ExcaliburGraphicsContextWebGL.ts:78]: Missing JSDoc warning like `__gl`?

File: src/engine/Graphics/Context/ExcaliburGraphicsContextWebGL.ts [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
+  private _transform = new TransformStack();
+  private _state = new StateStack();
+  private _ortho!: Matrix;
+
+  public __pointRenderer: PointRenderer;
+
+  public __lineRenderer: LineRenderer;
+
+  public __imageRenderer: ImageRenderer;
```
Comment (kamranayub) [inline src/engine/Graphics/Context/ExcaliburGraphicsContextWebGL.ts:88]: Do these need `@internal` JSDocs?

File: src/engine/Graphics/Context/draw-image-command.ts [hunk 10]
```diff
@@ -0,0 +1,141 @@
+import { Matrix } from '../../Math/matrix';
```
Comment (kamranayub) [inline src/engine/Graphics/Context/draw-image-command.ts:1]: Wonder if these classes/modules could benefit from comments while it's still fresh in your mind.

File: sandbox/src/game.ts (hunk 1 above)
Comment (eonarheim) [reply sandbox/src/game.ts]: It is not

File: sandbox/src/game.ts (hunk 2 above)
Comment (eonarheim) [reply sandbox/src/game.ts:505]: I think I like `.use` the most

File: src/engine/EntityComponentSystem/Components/TransformComponent.ts (hunk 5 above)
Comment (eonarheim) [reply src/engine/EntityComponentSystem/Components/TransformComponent.ts:21]: No, I think this was leftover from an experiment

File: src/engine/EntityComponentSystem/SystemManager.ts (hunk 6 above)
Comment (eonarheim) [reply src/engine/EntityComponentSystem/SystemManager.ts:68]: Oh the initialize lifecycle wasn't happening in certain cases, that's what's going on. We could type guard this as anything that has an engine? Or maybe specifically that context is a Scene?

**Pull request merged**